buildSrc/subprojects/packaging/src/main/kotlin/org/gradle/gradlebuild/packaging/ShadedJarPlugin.kt

```diff
@@ -0,0 +1,184 @@
+/*
+ * Copyright 2018 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.gradlebuild.packaging
+
+import accessors.base
+import org.gradle.api.Action
+import org.gradle.api.ActionConfiguration
+import org.gradle.api.Plugin
+import org.gradle.api.Project
+import org.gradle.api.artifacts.Configuration
+import org.gradle.api.artifacts.dsl.DependencyHandler
+import org.gradle.api.artifacts.transform.ArtifactTransform
+import org.gradle.api.attributes.Usage
+import org.gradle.api.file.ProjectLayout
+import org.gradle.api.model.ObjectFactory
+import org.gradle.api.tasks.bundling.Jar
+import org.gradle.kotlin.dsl.*
+
+
+private
+const val relocatedClassesAndAnalysisType = "relocatedClassesAndAnalysis"
+
+
+private
+const val relocatedClassesType = "relocatedClasses"
+
+
+private
+const val entryPointsType = "entryPoints"
+
+
+private
+const val classTreesType = "classTrees"
+
+
+private
+const val manifestsType = "manifests"
+
+
+/**
+ * Creates a shaded jar of the publication of the current project.
+ *
+ * The shaded jar is added as an artifact to the {@code publishRuntime} configuration.
+ */
+open class ShadedJarPlugin : Plugin<Project> {
+
+    override fun apply(project: Project): Unit = project.run {
+        val shadedJarExtension = createShadedJarExtension(createConfigurationToShade())
+
+        registerTransforms(shadedJarExtension)
+
+        val shadedJarTask = addShadedJarTask(shadedJarExtension)
+
+        plugins.withId("gradlebuild.publish-public-libraries") {
+            addShadedJarArtifact(shadedJarTask)
+        }
+    }
+
+    private
+    fun Project.createShadedJarExtension(configurationToShade: Configuration) =
+        extensions.create<ShadedJarExtension>("shadedJar", layout, objects, configurationToShade)
+
+    private
+    fun Project.registerTransforms(shadedJarExtension: ShadedJarExtension) {
+        afterEvaluate {
+            dependencies {
+                registerTransform {
+                    from
+                        .attribute(artifactType, "jar")
+                        .attribute(minified, true)
```

Comment:
This seems like a red flag, shouldn't these transforms be independent of one another? Are we missing some concept to tell Gradle "here's the order of transforms"?

Right now I wouldn't know how to hook in a step between minify and shade without changing the shade plugin.

Reply:
You are saying I shouldn't care if the artifact is minfied or not for shading?

The shading transform does not at all depend on the minification. It would even work the same no matter if the source is minified or not. Am I doing something wrong here?

Reply:
All I'm saying is that it's weird that there is any mention of the minify transform here. This might hint at a limitation in transforms or it might be an oversight. Just wanted to point it out.